File: Hevelius/others/workshop/UOS/Telescope/src/TelescopeImpl.cpp
/*******************************************************************************
 *    ALMA - Atacama Large Millimiter Array
 *    (c) Associated Universities Inc., 2002 *
 *    (c) European Southern Observatory, 2002
 *    Copyright by ESO (in the framework of the ALMA collaboration)
 *    and Cosylab 2002, All rights reserved
 *
 *    This library is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 2.1 of the License, or (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with this library; if not, write to the Free Software
 *    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
 *
 *
 *
 * "@(#) $Id: TelescopeImpl.cpp,v 1.10 2007/07/11 21:09:13 wg3 Exp $"
 * 
 * Implementation file developed by
 * Rodrigo Araya - rodrigo.araya@gmail.com
 * Sebastian Caro - carosebastian@gmail.com
 * Alejandro Barrientos - drlightspeed@msn.com
 *
 */

#include <TelescopeImpl.h>
#include <iostream>
#include <unistd.h>

#define __RCX_COMMAND_DELAYTIME__	50000   // delay time in u_sec between RCX command invocations 

using namespace std;

/* ----------------------------------------------------------------*/
TelescopeImpl::TelescopeImpl(const ACE_CString &name, maci::ContainerServices *containerServices) :
  acscomponent::ACSComponentImpl(name, containerServices)
{
  const char *_METHOD_="Telescope::TelescopeImpl";
  // ACS_TRACE is used for debugging purposes
  ACS_TRACE(_METHOD_);
}
/* ----------------------------------------------------------------*/
TelescopeImpl::~TelescopeImpl()
{
  const char *_METHOD_="Telescope::~TelescopeImpl";
  // ACS_TRACE is used for debugging purposes
  ACS_TRACE(_METHOD_);
  ACS_DEBUG_PARAM(_METHOD_, "Destroying %s...", name());
}

void
TelescopeImpl::execute()
  throw (ACSErr::ACSbaseExImpl)
{
  const char *_METHOD_="Telescope::execute";
  ACS_TRACE(_METHOD_);
  
    // Use container to activate the door object
   ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_DEBUG, "Getting LegoComponent component."));
  
  //  Working Lego Stuff.
  m_lego_p = H3E::LegoControl::_nil();
  
  m_lego_p = getContainerServices()->getDefaultComponent<H3E::LegoControl>("IDL:alma/H3E/LegoControl:1.0");
  
  if (CORBA::is_nil(m_lego_p.in()))
    {
      throw acsErrTypeLifeCycle::LifeCycleExImpl(__FILE__, __LINE__, "::Telescope::execute");
    }
  
  
  // Get Instrument Component Reference Here.
  m_Instrument_p = INSTRUMENT_MODULE::Instrument::_nil();
  m_Instrument_p = getContainerServices()->getDefaultComponent<INSTRUMENT_MODULE::Instrument>("IDL:alma/INSTRUMENT_MODULE/Instrument:1.0");
  
  if (CORBA::is_nil(m_Instrument_p.in()))
    {
      throw acsErrTypeLifeCycle::LifeCycleExImpl(__FILE__, __LINE__, "::Telescope::execute");
    }
  
}



/* --------------------- [ CORBA interface ] ----------------------*/

TYPES::ImageType *
TelescopeImpl::observe (const TYPES::Position & coordinates, CORBA::Long exposureTime)
  throw(CORBA::SystemException, UOSErr::PositionOutOfLimitsEx)
{
  const char *_METHOD_="Telescope::observe";
  
  ACS_TRACE(_METHOD_);
  ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_DEBUG, "Received observe command. coordinates:  , exposureTime: %f ",exposureTime));  
  
  
  /* Verify if the coordinates are within allowed ranges
   * Azimuth     0 - 360
   * Elevation  10 - 90
   */
  
  if ( (coordinates.az < 0) || (coordinates.az > 360) || (coordinates.el < 10) || (coordinates.el > 90)){
    
    ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_ERROR, "Coordinates are out of limits, allowed ranges are el: 10 - 90 , az: 0 - 360."));
    UOSErr::PositionOutOfLimitsExImpl ex(__FILE__,__LINE__,_METHOD_);
    ex.addData("Reason","Coordinates are out of limits.");
    throw ex.getPositionOutOfLimitsEx();
  }
  
  
  
  /*   This method send the setTo command to the Lego component and once in position sends the takeImage command
   *   to the Intrument component. Then returns the obtauined image.
   */
  TYPES::Position currentCoordinates = getCurrentPosition();
  TYPES::ImageType * picture;
  
  
  TelescopeImpl::moveTo(coordinates);
  
  // This variables defines the encoders errors in degrees
  double az_delta = 4;
  double el_delta = 4;
  
  
  
  
  ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_DEBUG, "Waiting for telescope to get in position az: %f, el: %f",coordinates.az, coordinates.el));
  
  
  while( ((currentCoordinates.az - coordinates.az) >  az_delta) ||  \
	 ((currentCoordinates.az - coordinates.az) < -az_delta) ||  \
	 ((currentCoordinates.el - coordinates.el) >  el_delta) ||  \
	 ((currentCoordinates.el - coordinates.el) < -el_delta)){
    
    // waits until the telescope is in position
    usleep(__RCX_COMMAND_DELAYTIME__);
    currentCoordinates = getCurrentPosition();
  }
  
  ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_DEBUG, "Telescope in position az: %f, el: %f",currentCoordinates.az, currentCoordinates.el));
  
  
  // Take the picture
  picture = m_Instrument_p->takeImage (exposureTime);
  ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_DEBUG, "Returning the taken image")); 
  return picture;
}


void
TelescopeImpl::moveTo(const TYPES::Position & coordinates)
  throw(CORBA::SystemException, UOSErr::PositionOutOfLimitsEx)
{
  
  const char *_METHOD_="Telescope::moveTo";
  
  
  ACS_TRACE(_METHOD_);
  ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_DEBUG, "Received moveTo command. coordinates az: %f el: %f", coordinates.az, coordinates.el));
  
  /* Verify if the coordinates are within allowed ranges
   * Azimuth     0 - 360
   * Elevation  10 - 90
   */
  
  if ( (coordinates.az < 0) || (coordinates.az > 360) || (coordinates.el < 10) || (coordinates.el > 90)){
    ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_ERROR, "Coordinates are out of limits, allowed ranges are el: 10 - 90 , az: 0 - 360."));
    UOSErr::PositionOutOfLimitsExImpl ex(__FILE__,__LINE__,_METHOD_);
    ex.addData("Reason","Coordinates are out of limits.");
    throw ex.getPositionOutOfLimitsEx();
  }
  
  
  // Create the callback for the objfix method
  //AsyncMethodCBvoid setToCB("setTo");
  ACS::CBvoid_var setTo_CB;
  ACS::CBDescIn setToDescIn;
  
  CORBA::Double altitude = coordinates.el;
  CORBA::Double azimuth = coordinates.az;
  
  ACS_LOG(LM_SOURCE_INFO,_METHOD_,(LM_DEBUG, "Trying to set coordinates az: %f, el: %f", azimuth, altitude));
  m_lego_p->setTo( altitude, azimuth, setTo_CB.in(), setToDescIn);
}


TYPES::Position
TelescopeImpl::getCurrentPosition()
  throw(CORBA::SystemException)
{
  
  const char *_METHOD_="Telescope::getCurrentPosition";
  ACS_TRACE(_METHOD_);
  ACS_LOG(LM_SOURCE_INFO,_METHOD_, (LM_DEBUG, "Received getCurrentPosition command."));
  
  TYPES::Position currentCoordinates;
  ACSErr::Completion *comp = new ACSErr::Completion();
  
  
  currentCoordinates.az = (CORBA::Double)m_lego_p->actualAzimuth()->get_sync(comp);
  usleep(__RCX_COMMAND_DELAYTIME__);
  currentCoordinates.el = (CORBA::Double)m_lego_p->actualAltitude()->get_sync(comp);
  
  return currentCoordinates;
}

/* --------------- [ MACI DLL support functions ] -----------------*/
#include <maciACSComponentDefines.h>
MACI_DLL_SUPPORT_FUNCTIONS(TelescopeImpl)
  /* ----------------------------------------------------------------*/
  
  
  /*___oOo___*/

File: Hevelius/src/trunk/amTCS/src/alma/TELESCOPE_MODULE/TelescopeImpl/TelescopeImpl.java
/*
 *    ALMA - Atacama Large Millimiter Array
 *    (c) European Southern Observatory, 2002
 *    Copyright by ESO (in the framework of the ALMA collaboration),
 *    All rights reserved
 *
 *    This library is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 2.1 of the License, or (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with this library; if not, write to the Free Software
 *    Foundation, Inc., 59 Temple Place, Suite 330, Boston, 
 *    MA 02111-1307  USA
 */

package alma.TELESCOPE_MODULE.TelescopeImpl;

import java.util.logging.Logger;

import alma.ACS.*;
import alma.TYPES.*;
import alma.acs.component.ComponentLifecycle;
import alma.acs.container.ContainerServices;
import alma.TELESCOPE_MODULE.TelescopeOperations;
import alma.TELESCOPE_MODULE.TelescopeImpl.TelescopeImpl;
import alma.acs.exceptions.AcsJException;
import alma.acs.component.ComponentLifecycleException;
//import si.ijs.maci.ComponentInfo;
import alma.acs.component.ComponentDescriptor;
import alma.JavaContainerError.wrappers.AcsJContainerServicesEx;


public class TelescopeImpl implements TelescopeOperations, ComponentLifecycle {

	private ContainerServices m_containerServices;
	private Logger m_logger;
	private boolean working;
	private boolean stopping;

	private alma.TELESCOPE_MODULE.Telescope tele_comp;
	private alma.TRACKING_MODULE.Tracking trck_comp;
	private TelescopeWorker worker;

	org.omg.CORBA.Object obj = null;

	/////////////////////////////////////////////////////////////
	// Implementation of ComponentLifecycle
	/////////////////////////////////////////////////////////////
	
	public void initialize(ContainerServices containerServices) throws ComponentLifecycleException {
		m_containerServices = containerServices;
		m_logger = m_containerServices.getLogger();
                working = false;
                stopping = false;
		m_logger.info("initialize() called...");
		try {
			obj = m_containerServices.getDefaultComponent("IDL:alma/TRACKING_MODULE/Tracking:1.0");
			trck_comp = alma.TRACKING_MODULE.TrackingHelper.narrow(obj);
		}
		catch (alma.JavaContainerError.wrappers.AcsJContainerServicesEx e) {
			m_logger.fine("Failed to get Tracking component reference " + e);
			throw new ComponentLifecycleException("Failed to get Tracking component reference");
		}
	}

	public void execute() throws ComponentLifecycleException {
		m_logger.info("execute() called...");
	}

	public void cleanUp() {
		m_logger.info("cleanUp() called..., nothing to clean up.");
	}

	public void aboutToAbort() {
		if(working)
		{
			m_logger.fine("Trying to stop");
			//		try
			//		{
			stop();
			//		}
			//		catch(SchedulerAlreadyStoppedEx e)
			//		{
			//			m_logger.fine("Already Stopped");
			//		}
		}
		cleanUp();
		m_logger.info("managed to abort...");
	}

	/////////////////////////////////////////////////////////////
	// Implementation of ACSComponent
	/////////////////////////////////////////////////////////////


	public ComponentStates componentState() {
		return m_containerServices.getComponentStateManager().getCurrentState();
	}
	public String name() {
		return m_containerServices.getName();
	}

	/////////////////////////////////////////////////////////////
	// Implementation of TelescopeOperations
	/////////////////////////////////////////////////////////////

	private double Alt = 0;
	private double Az = 0;
	private double Ra = 0;
	private double Dec = 0;
	private double ST = 0;

	public void goToAltAz(AltazPos p, AltazVel v, alma.ACS.CBvoid cb, alma.ACS.CBDescIn desc) {
		Alt = p.alt;
		Az = p.az;
		hor2radec(p);
	}

	public void getPos(RadecPosHolder p_rd, AltazPosHolder p_aa) {
		if(p_rd!=null)
		{
			p_rd.value = new RadecPos();
			p_rd.value.ra = Ra;
			p_rd.value.dec = Dec;
		}
		if(p_aa!=null)
		{
			p_aa.value = new AltazPos();
			p_aa.value.alt = Alt;
			p_aa.value.az = Az;
		}
		System.out.println("Llegue Tel Asdas by Mex");
		//System.out.println(p_aa.value.alt+" "+p_aa.value.az+" "+p_rd.value.ra+" "+p_rd.value.dec);
	}

	public void hor2radec(alma.TYPES.AltazPos p)
	{
		TelescopeConverter.altaz2radec(p.alt, p.az);
		Ra = TelescopeConverter.getRa();
		Dec = TelescopeConverter.getDec();
		ST = TelescopeConverter.getST();
	}

	public void radec2hor(alma.TYPES.RadecPos p)
	{
		TelescopeConverter.radec2altaz(p.ra, p.dec);
		Alt = TelescopeConverter.getAlt();
		Az = TelescopeConverter.getAz();
		ST = TelescopeConverter.getST();
	}

	public void preset(alma.TYPES.RadecPos p, alma.ACS.CBvoid cb, alma.ACS.CBDescIn desc){

		Ra = p.ra;
		Dec = p.dec;
		radec2hor(p);

	}

	public double getSideralTime()
	{
		return ST;
	}

	public boolean isWorking()
	{	
		return working;
	}

	public void setWorking(boolean s)
	{
		working = s;
	}

	public void start() {//throws AcsJException{
		m_logger.info("Start called");

		if (!working && !stopping) {
			working = true;
/*
                try {
                        obj = m_containerServices.getDefaultComponent("IDL:alma/TELESCOPE_MODULE/Telescope:1.0");
                        tele_comp = alma.TELESCOPE_MODULE.TelescopeHelper.narrow(obj);
                } catch (alma.JavaContainerError.wrappers.AcsJContainerServicesEx e) {
                        m_logger.fine("Failed to get Telescope component reference " + e);
                        throw new ComponentLifecycleException("Failed to get Telescope component reference");
                }
*/
                try
                {
                        ComponentDescriptor s = m_containerServices.getComponentDescriptor("TELESCOPE");
                        obj = s.getComponent();
                        tele_comp = alma.TELESCOPE_MODULE.TelescopeHelper.narrow(obj);
                }
                catch(AcsJContainerServicesEx e)
                {
                        m_logger.fine("da");
                }

			worker = new TelescopeWorker(m_logger, trck_comp, tele_comp);
			worker.start();
		} else {
			m_logger.info("Already working");
			//AcsJException e = new AcsJException();
			//throw e;
		}
	}

	public void stop() {//throws AcsJException {
		m_logger.info("Stop called");

		if (!stopping)
			stopping = true;
		else {
			m_logger.info("Already trying to stop");

			//TODO: perhaps throw another exception here...
			return;
		}

		if (!working) {
			m_logger.info("Not yet working");
			stopping = false;
			//AcsJException e = new AcsJException();
			//throw e;
		}
		working = false;

		//block until last proposal finished
		if (worker != null) {
			try {
				m_logger.fine("Waiting for thread to die...");
				worker.join();
			} catch(Exception e) {
				m_logger.fine("Waiting for thread to die error");
			}
		}

		worker = null;
		stopping = false;

		m_logger.fine("Thread terminated");
	}
/*
	protected setRaDec(RadecPos p)
	{
		Ra = p.ra;
		Dec = p.dec;
	}

	protected setAltAz(AltazPos p)
	{
		Alt = p.alt;
		Az = p.az;
	}
*/
}
